### ISSUE TITLE
Caching & More Robust Loading

### ISSUE THREAD

**Issue by dcvz** [jorgenhenrichsen/SwiftAudio#39]
Wanted to bring this to your attention: https://github.com/neekeetab/CachingPlayerItem

I've done some basic test integrating into the library and it's interesting.

I think a track should play ASAP - and caching in queued mode should automatically download the next track when the current one is fully loaded.

This doesn't provide all of that but it has some nice ideas.

**Comment by jorgenhenrichsen** [comment]
Nice!

Yeah, I have thought about the same thing and it is definitely something that should be added 👍

**Comment by rawandahmad698** [comment]
@jorgenhenrichsen; YES, this is  required. Add an ability to download and cache an item, most of my songs are in mp3 and 7>10 MB, so this won't hurt. I do truly love your work. It's clean!

**Comment by murad1981** [comment]
yes the library is so clean, a small suggestion is to use the mentioned custom AVPlayerItem in SwiftAudio.

**Comment by JhonnyTawk** [comment]
@dcvz Hey, I am trying to add caching to SwiftAudio using https://github.com/neekeetab/CachingPlayerItem

but the func add and load take AudioItem `add(items: [AudioItem], playWhenReady: Bool = true)`

and in the caching library AVPlayer take CachingPlayerItem so how to integrate it?

Thank in Advance

**Comment by nishanBende** [comment]
@dcvz We could also add a preloading functionality. 
Ideas:
https://github.com/delannoyk/AudioPlayer/issues/36#issuecomment-198806154

Thanks @dcvz for react-native-track-player. I came looking if the underlying SwiftAudio provides a preloading functionality. 
I am not an iOS developer but will try to implement this feature in a fork. I'll let you know if I come up with something.
This is a must-have to built apps with awesome feeds (Instagram e.g.) with almost 0 load time.

**Comment by limelightapp** [comment]
@nishanBende let me know when you do! That would be an amazing feature.

**Comment by nishanBende** [comment]
@dcvz I have attempted a preload functionality in SwiftAudio and tested with react-native-track-player. 
The playing becomes as fast as a downloaded asset. I have been testing this for a few days.
https://github.com/nishanBende/SwiftAudio/commit/41e5564c4f604c86a5c8802eab627c887c2b8d80

P.S. I am not sure about the implications it can bring, I took the implementation idea from this comment.
https://github.com/delannoyk/AudioPlayer/issues/36#issuecomment-198806154

**Comment by nishanBende** [comment]
@dcvz I am managing the queue in JavaScript because the underneath QueueAudioPlayer is not preloading the next track so I have exported the preload function to RNTrackPlayer with preload/cancelPreload implementation in this commit.
https://github.com/nishanBende/SwiftAudio/commit/41e5564c4f604c86a5c8802eab627c887c2b8d80

I feel we should split it into 2 separate features.
1. Caching while Playing. Similar to  https://github.com/neekeetab/CachingPlayerItem.
2. Preloading any asset programmatically.

Can we also let AVQueuePlayer handle the queue for us? 
There's some discussion that it internally preloads the next track but I cannot confirm.
https://developer.apple.com/documentation/avfoundation/avqueueplayer

**Comment by whsmacon** [comment]
I also come from react-native-track-player and would love to see gapless queues implemented.
